### PR TITLE
Have an nice looking bash prompt

### PR DESCRIPTION
**Pull request by sunix**
### What does this PR do?
Replace the default bash prompt with one containing informations about the Che container:
- CHE_WORKSPACE_NAMESPACE
- CHE_WORKSPACE_NAME
- CHE_MACHINE_NAME (the component/container name)
- current directory

### Screenshot/screencast of this PR
Selection_049

### What issues does this PR fix or reference?
no reference issues

### How to test this PR?
- start the quarkus get started stack
- open a terminal in the `centos-quarkus-maven` container
- run the command: 
  ```.bashrc
  echo "PS1='\[\e[38;5;69m\]$CHE_WORKSPACE_NAMESPACE\[\e[0;39m\]@\[\e[38;5;220m\]$CHE_WORKSPACE_NAME\[\e[0;39m\] 🟢 $CHE_MACHINE_NAME\[\e[0;39m\]:\[\e[38;5;172m\]\w\[\e[0;39m\]\[\e[1;32m\]\[\e[0;39m\] \n\[\e[38;5;172m\]>_ \[\e[1;39m\]'" > "${HOME}"/.bashrc
  ```

- open a new terminal
- see the new prompt

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by benoitf** [comment]
Enhancing some prompts might be good but then it won't be consistent across all sidecars
I'm not sure to see the value of 🟢  ? I mean, I never saw that on prompts

**Comment by vitaliy-guliy** [comment]
What does green circle mean? Is it just an icon indicating running container?

**Comment by sunix** [comment]
the green circle is a recall to the one in My workspace view

**Comment by benoitf** [comment]
I would say that I like to have some workspace name + machine name info but
I don't like to start over the next line(it reduces every lines by 2), the namespace (99% I don't care) and the green circle

**Comment by sunix** [comment]
Selection_051

**Comment by benoitf** [comment]
ideally to me, prompt is quite short, it should not take half of the width of my screen

**Comment by sunix** [comment]
> Enhancing some prompts might be good but then it won't be consistent across all sidecars

if we could have consistent prompt for the default "user runtimes" containers it is fair enough.

**Comment by benoitf** [comment]
Some information is also redundant with the terminal title
In your screenshot, then I don't see why `centos-quarkus-maven` is duplicated between the title and the prompt for example

**Comment by sunix** [comment]
@benoitf right but somehow people can't see it. what is your proposal ?

**Comment by sunix** [comment]
I prefer having a new line so i can have in 1 line the maximum characters for my command where ever the folder where I am.
But these are personal preferences, at some point we have to agree to something that some people would like and others won't.

**Comment by vitaliy-guliy** [comment]
I also prefer to have a shorter prompt and without adding a new line.

**Comment by benoitf** [comment]
I think it's confusing to have all this information in the prompt. Because here we're always connected to the local workspace

machine-name: no need to put in the prompt (it's in the title)
workspace-name: no need to put in the prompt (It's in the URL but agreed that it could go inside the IDE but then probably inside the status bar and not on all terminal prompts as we're always on the same workspace)
namespace: should go in status bar as it's common to all widgets
The circle: it's quite lying and does not bring any value. Also if theia disconnects the prompt will stay with green color while I can't use it anymore.

On my local desktop I like to have the current git branch , if some files are missing, etc with zsh but inside the IDE you've the branch name displayed in the status bar so I don't need it there.

so basically I only really like to have "colors" on `/projects`

If I look at some other cloud IDE like gitpod and codespace:
image

Screenshot 2020-11-17 at 16 39 20

it's `project name` and `current folder`

**Comment by sunix** [comment]
We have to take advantage on the fact that we are in a pod with possibly multiple containers. It is important that the user understand it and get the value because this is how Che is different. I prefer people to be confuse rather than not seeing the difference with a traditional local terminal or cloud IDE terminal.

**Comment by azatsarynnyy** [comment]
my 2 cents:

I think with these changes command line prompt becomes too overloaded (with the duplicated information BTW) and it can eat too much space.
To me, command line prompt has to contain minimum context information that's really useful, and most of the time it changes in different containers, like the current folder or git branch.

_Workspace's namespace and name_ are already displayed in the browser's navigation bar.
Usually, the user is aware of in what namespaces their Workspace is running. If not - maybe, one just doesn't need that info? Like most of the users?
And, I guess, most of the users are operating in a single namespace.

For a _Workspace name_, my opinion is the same as for a namespace.

If we look at one of the real examples when I'm checking the PR, clicking on `Try it` button from che-bot, the `namespace/ws-name` is really long, e.g. `azatsary/pr-che_theia-nodejs-908-7e0b6`
It would eat too much space on my command line prompt, duplicating the information from the browser's navigation bar.

A _container name_ is duplicated. It already takes half of the Terminal's title bar. So, why waste more space on the command line prompt for displaying exactly the same information?

Doesn't green ball is confusing in command line prompt for anyone trying to understand what does that mean?

**Comment by sunix** [comment]
This is not fair...
You are all saying that you want to save space but at the same time you are OK to display the git branch in the prompt even if it is display in the status bar.
How many times have you started typing a command in a terminal ... and figure out few minutes later that you are not in the right one even if the name of the container is in the tile. It happens all the time.

Like the screenshots I have provided were not from a real case projects.... it is long but I can fit 2 terminals in the same workspace window.

Selection_052

Even if some of you may not like it, we should merge that PR, some others would like it and it is definitely better than before.

**Comment by azatsarynnyy** [comment]
> ... you are OK to display the git branch in the prompt even if it is display in the status bar.

agree, git branch is redundant info in command line prompt

**Comment by vzhukovs** [comment]
I would support @benoitf quote around displaying only working directory in prompt. As for me, terminal prompt should be as simpler as possible. Without any extra information.

e.g. my terminal prompt:
vlad@MBP-15-Vlad: ~ 2020-11-17 21-31-38

**Comment by sunix** [comment]
if redundant info is helping the UX, then I don't see any reason not to have it. For the container name, it helps me at least.

**Comment by sunix** [comment]
This won't break your development flow and won't make you less productive. The code is OK (there is no code).
If there is no objection except personal preferences, please approve this PR.

**Comment by benoitf** [comment]
I think that what you could do is to make: minimal changes on what people here seems to agree (current directory with colors)

That said, you could bring an editor preference feature allowing to customize .bashrc, or something like that for all terminals based on sh/bash for example

Then you could have what you're suggesting, even proposing different pre-defined prompts but IMO what you're suggesting can't really be the default based on previous feedbacks.

**Comment by RomanNikitenko** [comment]
> That said, you could bring an editor preference feature allowing to customize .bashrc, or something like that for all terminals based on sh/bash for example

+1 to have some mechanism which allows to customize terminals as a user prefer

**Comment by sunix** [comment]
Some others seem to like it on twitter: https://twitter.com/__sunix_/status/1328333226977406976 and https://twitter.com/sebi2706/status/1328727818734366727

**Comment by sunix** [comment]
@benoitf
Your screenshot with gitpod terminal is displaying "gitpod". This is redundant info.
Your screenshot with codespace terminal is displaying "codespace". This is redundant info.
I don't think anybody are complaining.

**Comment by sunix** [comment]
If you have other suggestion, feel free to open a PR. And we could make a public poll and see what people like or not.

**Comment by sunix** [comment]
I can completely understand the minimal mindset for a traditional terminal in your local laptop. OK to make a way to customize and use my custom prompt. But at this stage, Che and the idea of having a cloud workspace with multiple containers in a Kubernetes pod is not a well known. Anything that shows a bit of that should help new user to figure out how good and different Che could be compared to traditional IDE/editors or competitor Cloud IDEs.

**Comment by sympatheticmoose** [comment]
@sunix thanks for raising this idea :) so I think the issue you are trying to highlight you've started to mention in that last comment - 

> Che and the idea of having a cloud workspace with multiple containers in a Kubernetes pod is not a well known.

and related to this is the problem
> How many times have you started typing a command in a terminal ... and figure out few minutes later that you are not in the right one even if the name of the container is in the tile. It happens all the time.

That is probably a valid concern, and we should explore how we can better help users understand this in Che. However, there is not consensus that this approach makes it a better experience or actually solves that problem fully, and the only folks who have actually commented on the issue prefer a more minimal approach within the terminal prompt. It's good that you've put this on twitter to try and get some more feedback, lets see if we actually get some, I don't think we can count a small amount of likes as enough to proceed.

What @benoitf suggests around an editor preference feature is a nice idea to help cater for multiple preferences, but I do not believe there is actual evidence that this is inherently better, and would be reluctant for this to become the **default** based on the discussion here.

**Comment by vzhukovs** [comment]
`container_name path $` prompt looks better imo, for example:

```
che-quarkus-maven ~ $
che-quarkus-maven /projects $
che-quarkus-maven / $
che-quarkus-maven /tmp $
```

**Comment by sunix** [comment]
Thanks @sympatheticmoose 

About twitter, I don't have enough followers ;), we could create a poll with the Che account. Now how would you phrase it?

So this prompt provides the relevant information whenever the context: whether you are in terminal full screen, a terminal shown in a tab or the same terminal from the Openshift console:

It shows where you are:
- a folder, located in ...
- a container, located in ...
- a pod/workspace, located in ...
- a user namespace

**Comment by ericwill** [comment]
Bash configs and other improvements are often a highly subjective topic, with everyone having a different opinion of what's right. This thread is good evidence of that.

IMO the enhancement here lies not in hard coding a certain prompt style, but by letting the user configure prompts to their liking. Something like a user preference or config file that Che picks up on and uses would be good. I understand the want to hard code a certain style, and while that's nice and good now, it isn't sustainable when someone comes with a new PR wanting to change it again. Twitter polls and feedback for every such change are also not sustainable. This is one of those areas where we should give the user the tool to customize their environment, but not actually do it for them.

@sunix I would support working on a way to allow the user to provide their own bash configs, either through a file or some other setting. Hard coding the prompt style isn't the correct way forward IMO.

**Comment by vitaliy-guliy** [comment]
I'm not sure how created pool can help us. The people, that give +1 could be not Che-Theia users, but vote +1 just because the terminal looks more colored / more fun.

But, having a separate section in preferences to be able to configure parts of prompt (your comment above) would be nice
- [ ] a folder, located in ...
- [ ] a container, located in ...
- [ ] a pod/workspace, located in ...
- [ ] a user namespace
- [ ] a git branch ??

**Comment by scela** [comment]
Let's not do Twitter-based development please.

**Comment by sunix** [comment]
Let's do Che first timer based development.

I am fine to provide a way to customize it, but we need to provide a default one anyway. The prompt should aim Che first timer developers. 

Many things are subjective for developers: colors, shortcuts, layout, extensions. What if by default, we provide no syntax coloration, we do not display the line numbers, etc ... At some point we should not take decision based on personal taste.
It should be done base on a goal:
- Provide the minimal to any user to use the product at its best
- Allow customisation and configuration to use the platform at its best

The goal of this prompt is to provide context information when using the terminal:
> how many times have you started typing a command in a terminal ... and figure out few minutes later that you are not in the right one even if the name of the container is in the tile. It happens all the time.

All the points against are quite subjective:
> I also prefer to have a shorter prompt and without adding a new line.

or

> I think with these changes command line prompt becomes too overloaded (with the duplicated information BTW) and it can eat too much space.

**Comment by scela** [comment]
No one here is advocating to not make it easy for people to use the products. It's a matter of how. Having green balls around is your personal taste. According to my taste it makes the terminal look like a Christmas tree.

**Comment by sunix** [comment]
`git terminal 17` without the christmas tree:

Selection_053

**Comment by sunix** [comment]
It is not my taste, I just try reuse the same icon we have in the `My workspace` view:

Selection_051

**Comment by vzhukovs** [comment]
What can I suggest, to resolve this situation is to collect all possible options, e.g.:

- your variant
- another option
- one more option
- leave it as is

then create the poll in google forms and send it to che-dev and show the result of voting after one week, then it will be fair deal for everyone.

**Comment by sunix** [comment]
I think we want the best for our user and futur user, so if we go for a pool I would also include:

- solution architects: they are testing and making demos based on our products to our customers
- developer advocates: they are testing and making demos based on our community projects to any developers
- product managers

Based on the pool and any other aspect, decision maker should decide: should be David or Mario for our group.

**Comment by sunix** [comment]
new proposal: 
- a welcome message with infos
- 1 line prompt
- just the container name + folder in the prompt

Selection_064

**Comment by gattytto** [comment]
this could include more functions for the kubernetes sidecar because kubectl and tkn and others come with autocompletion templates that we can add to bash or zsh.

**Comment by ericwill** [comment]
There doesn't seem to be consensus on this PR so I'm going to close it. Maybe we can discuss this topic at a community call and take it from there.
